src/Common/src/TypeSystem/Common/MetadataTypeSystemContext.cs

```diff
@@ -61,23 +78,33 @@ public virtual void SetSystemModule(ModuleDesc systemModule)
             InitializeSystemModule(systemModule);
 
             // Sanity check the name table
-            Debug.Assert(s_wellKnownTypeNames[(int)WellKnownType.MulticastDelegate - 1] == "MulticastDelegate");
+            Debug.Assert(s_wellKnownTypes[(int)WellKnownType.MulticastDelegate - 1].Name == "MulticastDelegate");
 
-            _wellKnownTypes = new MetadataType[s_wellKnownTypeNames.Length];
+            _wellKnownTypes = new MetadataType[s_wellKnownTypes.Length];
 
             // Initialize all well known types - it will save us from checking the name for each loaded type
             for (int typeIndex = 0; typeIndex < _wellKnownTypes.Length; typeIndex++)
             {
-                MetadataType type = systemModule.GetType("System", s_wellKnownTypeNames[typeIndex]);
-                type.SetWellKnownType((WellKnownType)(typeIndex + 1));
-                _wellKnownTypes[typeIndex] = type;
+                WellKnownTypeDef wellKnownType = s_wellKnownTypes[typeIndex];
+                MetadataType type = systemModule.GetType("System", wellKnownType.Name, !wellKnownType.IsOptional);
```

Comment:
`throwIfNotFound` argument can be just `wellKnownType == WellKnownType.Object` now. No need to have a separate column in the static array.

Also, it may be worth adding a comment to explain it. Something like: `Require System.Object to be present as a minimal sanity check. The set of required well-known types is not strictly defined since different .NET profiles implement different subsets.`